### DISCUSSION TITLE
[General] <withdraw contract's balance>

### DISCUSSION THREAD

**Discussion by FireLoveInIce** [zkSync-Community-Hub/zksync-developers#131] [General] [answered]
### Environment

Mainnet

### Provide a brief description of the functionality you're trying to implement.

I have defined the withdraw function using the following method, and now that the contract is deployed, I've noticed it's not working correctly on zksync. How can I withdraw the balance from the contract?
`function withdraw(address payable _to, uint256 _amount) external onlyOwner {
        require(_to != address(0), "Invalid address");
        require(
            _amount > 0 && _amount <= address(this).balance,
            "Invalid amount"
        );
        _to.transfer(_amount);
    }`

### What is the specific issue or error you're encountering?

nothing, here is the failed withdraw transaction:https://explorer.zksync.io/tx/0x96b5058c00712b3aef2beb97449c4c8f118f5f0cdc941e76b8c3e468218f9bc5

### Can you share the error messages or logs you're receiving, if any?

{
  "transactionHash": "0x96b5058c00712b3aef2beb97449c4c8f118f5f0cdc941e76b8c3e468218f9bc5",
  "transactionIndex": 0,
  "blockHash": "0x53c046fd6c0e6981810f6665ad6e1a8e0c6186100cdf37c9b8aa0e8023ec50e2",
  "blockNumber": 13558812,
  "l1BatchTxIndex": null,
  "l1BatchNumber": null,
  "from": "0x841c4582c8a68aae3b660367c39a85b064d9f574",
  "to": "0xdf88d043aaea46f2e43d7a589d47eab0d024401d",
  "cumulativeGasUsed": 0,
  "gasUsed": 269405,
  "contractAddress": null,
  "l2ToL1Logs": [],
  "status": false,
  "root": "0x53c046fd6c0e6981810f6665ad6e1a8e0c6186100cdf37c9b8aa0e8023ec50e2",
  "logsBloom": "0x00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000",
  "type": "0x2",
  "effectiveGasPrice": "0xee6b280",
  "events": {
    "Transfer": [
      {
        "address": "0x000000000000000000000000000000000000800A",
        "blockHash": "0x53c046fd6c0e6981810f6665ad6e1a8e0c6186100cdf37c9b8aa0e8023ec50e2",
        "blockNumber": 13558812,
        "l1BatchNumber": null,
        "transactionHash": "0x96b5058c00712b3aef2beb97449c4c8f118f5f0cdc941e76b8c3e468218f9bc5",
        "transactionIndex": 0,
        "logIndex": 0,
        "transactionLogIndex": "0x0",
        "logType": null,
        "removed": false,
        "id": "log_5bddde62",
        "returnValues": {
          "0": "0x841c4582c8a68AAe3b660367c39A85b064d9F574",
          "1": "0x0000000000000000000000000000000000008001",
          "2": "804732500000000",
          "from": "0x841c4582c8a68AAe3b660367c39A85b064d9F574",
          "to": "0x0000000000000000000000000000000000008001",
          "value": "804732500000000"
        },
        "event": "Transfer",
        "signature": "0xddf252ad1be2c89b69c2b068fc378daa952ba7f163c4a11628f55a4df523b3ef",
        "raw": {
          "data": "0x0000000000000000000000000000000000000000000000000002dbe662cd3d00",
          "topics": [
            "0xddf252ad1be2c89b69c2b068fc378daa952ba7f163c4a11628f55a4df523b3ef",
            "0x000000000000000000000000841c4582c8a68aae3b660367c39a85b064d9f574",
            "0x0000000000000000000000000000000000000000000000000000000000008001"
          ]
        }
      },
      {
        "address": "0x000000000000000000000000000000000000800A",
        "blockHash": "0x53c046fd6c0e6981810f6665ad6e1a8e0c6186100cdf37c9b8aa0e8023ec50e2",
        "blockNumber": 13558812,
        "l1BatchNumber": null,
        "transactionHash": "0x96b5058c00712b3aef2beb97449c4c8f118f5f0cdc941e76b8c3e468218f9bc5",
        "transactionIndex": 0,
        "logIndex": 1,
        "transactionLogIndex": "0x1",
        "logType": null,
        "removed": false,
        "id": "log_62e9f6aa",
        "returnValues": {
          "0": "0x0000000000000000000000000000000000008001",
          "1": "0x841c4582c8a68AAe3b660367c39A85b064d9F574",
          "2": "737381250000000",
          "from": "0x0000000000000000000000000000000000008001",
          "to": "0x841c4582c8a68AAe3b660367c39A85b064d9F574",
          "value": "737381250000000"
        },
        "event": "Transfer",
        "signature": "0xddf252ad1be2c89b69c2b068fc378daa952ba7f163c4a11628f55a4df523b3ef",
        "raw": {
          "data": "0x00000000000000000000000000000000000000000000000000029ea4f3786480",
          "topics": [
            "0xddf252ad1be2c89b69c2b068fc378daa952ba7f163c4a11628f55a4df523b3ef",
            "0x0000000000000000000000000000000000000000000000000000000000008001",
            "0x000000000000000000000000841c4582c8a68aae3b660367c39a85b064d9f574"
          ]
        }
      }
    ]
  }

### Have you made any recent changes to the contract before encountering this issue?

no

### Are there any external libraries or contracts that your contract interacts with?

yes,my contract is extend openzeppelin's ERC20 and Ownable, but I think it isn't the point.

### Can you provide the relevant portions of your contract code where the issue is occurring?

`function withdraw(address payable _to, uint256 _amount) external onlyOwner {
        require(_to != address(0), "Invalid address");
        require(
            _amount > 0 && _amount <= address(this).balance,
            "Invalid amount"
        );
        _to.transfer(_amount);
    }`

### Have you tried to isolate the problem, and if so, what were the results?

not work.

### What steps have you already taken to try to resolve the issue?

no idea.

**Comment by vasyl-ivanchuk** [comment]
Hi @FireLoveInIce, your transaction has failed because of the following error: `Ownable: caller is not the owner`. The address that initiated the transaction (withdrawal) is not the owner of the contract and since `withdraw` uses `onlyOwner` from `Ownable.sol` 
```
function withdraw(address payable _to, uint256 _amount) external onlyOwner
```
only the owner is allowed to withdraw.

**Reply by FireLoveInIce** [reply]
I called the contract using the EOA (Externally Owned Account) of the owner. I have identified the issue, which is that zksync does not support the native Ethereum transfer function for transfers; it can only be done using the call function. Unfortunately, I cannot withdraw my balance from the contract, and it's really frustrating.

**Reply by vasyl-ivanchuk** [reply]
zkSync is  EVM-compatible and supports `transfer` function. If you want to transfer ERC20 token in the withdrawal function, check [this](https://stackoverflow.com/questions/68545930/how-to-withdraw-all-tokens-from-the-my-contract-in-solidity) StackOverflow discussion to see the difference in how ERC20 and ETH are transferred.

Your contract is built based on OpenZeppelin ERC20 implementation so [here](https://docs.openzeppelin.com/contracts/2.x/api/token/erc20#IERC20-transfer-address-uint256-) is the official doc page about ERC20 `transfer` function.

**Reply by FireLoveInIce** [reply]
<img width="593" alt="image" src="https://github.com/zkSync-Community-Hub/zkync-developers/assets/32660493/0ef7028c-f298-4398-ad9e-ff1960bf8c0d">

AiBot told me that.

I want to withdraw ETH balance in contract, and I'm sure the the function:
`function withdraw(address payable _to, uint256 _amount) external onlyOwner { require(_to != address(0), "Invalid address"); require( _amount > 0 && _amount <= address(this).balance, "Invalid amount" ); _to.transfer(_amount); }` 
is error cause the code:
 ` _to.transfer(_amount)`.

**Reply by vasyl-ivanchuk** [reply, marked as answer]
To my knowledge, native transfers are supported (yet not recommended) but with the gas limit specified for the transaction. Please, let me know if that works for you.

Here is a good blog post about why `transfer()` should not be used: [Stop Using Solidity's transfer() Now](https://consensys.io/diligence/blog/2019/09/stop-using-soliditys-transfer-now/). Please, also check our zkSync Era [Security and best practices ](https://era.zksync.io/docs/dev/building-on-zksync/best-practices.html) page where we also have a [section](https://era.zksync.io/docs/dev/building-on-zksync/best-practices.html#use-call-over-send-or-transfer) for this.

**Reply by FireLoveInIce** [reply]
yeah, I deploy a new contract which using call() instead of transfer(), it works well. But the ETH balance in old contract can not be withdrawn anymore.
